google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java

```diff
@@ -1490,6 +1490,41 @@ public void readOnlyTransaction_failureWhenAttemptingWrite()
     }
   }
 
+  @Test
+  public void readOnlyTransaction_successfulRead() throws Exception {
+    DocumentReference documentReference = randomColl.add(SINGLE_FIELD_MAP).get();
+
+    Timestamp firstWriteTime =
+        documentReference.set(Collections.singletonMap("counter", 1)).get().getUpdateTime();
+    documentReference.set(Collections.singletonMap("counter", 2)).get();
+
+    final TransactionOptions option =
+        TransactionOptions.createReadOnlyOptionsBuilder()
+            .setReadTime(
+                com.google.protobuf.Timestamp.newBuilder()
+                    .setSeconds(firstWriteTime.getSeconds())
+                    .setNanos(firstWriteTime.getNanos()))
+            .build();
+
+    final AtomicReference<DocumentSnapshot> ref = new AtomicReference<>();
+
+    final ApiFuture<Void> runTransaction =
+        firestore.runTransaction(
+            transaction -> {
+              final DocumentSnapshot snapshot =
+                  transaction.get(documentReference).get(5, TimeUnit.SECONDS);
+              ref.compareAndSet(null, snapshot);
+              return null;
```

Comment:
You could just return the value of the counter here. Then this transaction returns an `ApiFuture<Integer>` and you don't have to copy the document state.